### PR TITLE
Dispatcher improvements

### PR DESCRIPTION
**Pull request by geeknoid**
- Fix bug around handling of quotas. Mixer wasn't observing the requested quota name and getting quota from all defines instances. It now limits itself to the requested instances and produces an error if the proxy requests quota from an unknown instance, instead of failing open.

- Pass CheckResult, QuotaResult, and QuotaMethodArgs by value. This
is cleaner, avoids transient allocations, and eliminates some code paths.

### CODE REVIEW

File: mixer/pkg/runtime/dispatcher/session.go [hunk 1]
```diff
@@ -231,9 +232,8 @@ func (s *session) waitForDispatched() {
 			st = state.checkResult.Status
 
 		case tpb.TEMPLATE_VARIETY_QUOTA:
-			if s.quotaResult == nil {
-				r := state.quotaResult
-				s.quotaResult = &r
+			if status.IsOK(s.quotaResult.Status) && s.quotaResult.ValidDuration == 0 && s.quotaResult.Amount == 0 {
```
Comment (ozevren) [inline mixer/pkg/runtime/dispatcher/session.go]: Nit: add an IsDefault() / IsEmpty() method on QuotaResult?

Comment (kyessenov) [reply mixer/pkg/runtime/dispatcher/session.go]: +1. This also strikes me as begging to define "IsEmpty" method here and for `CheckResult`.

**Comment by geeknoid** [comment]
Addressed feedback, PTAL.

File: mixer/pkg/api/grpcServer.go [hunk 2]
```diff
@@ -161,17 +146,12 @@ func (s *grpcServer) check(legacyCtx legacyContext.Context, req *mixerpb.CheckRe
...
 			qr, err = s.dispatcher.Quota(legacyCtx, checkBag, qma)
 			if err != nil {
 				err = fmt.Errorf("performing quota alloc failed: %v", err)
 				lg.Errora("Quota failure:", err.Error())
 				// we continue the quota loop even after this error
-			} else if qr == nil {
-				// If qma.Quota does not apply to this request give the client what it asked for.
-				// Effectively the quota is unlimited.
-				crqr.ValidDuration = defaultValidDuration
-				crqr.GrantedAmount = qma.Amount
```
Comment (kyessenov) [inline mixer/pkg/api/grpcServer.go]: Can you please explain where this line is now? The default quota should be what was requested, not 0.

Comment (geeknoid) [reply mixer/pkg/api/grpcServer.go]: Well, this question triggered a bunch fo investigation. We have a bug in how we're handling quotas in Mixer at the moment. I'll have to work on that before submitting this change...

**Comment by geeknoid** [comment]
Fixes newly discovered quota bug.

PTAL.

File: mixer/pkg/runtime/dispatcher/session.go [hunk 3]
```diff
@@ -142,8 +143,18 @@ func (s *session) dispatch() error {
 			ndestinations++
 
 			for j, input := range group.Builders {
+				if s.variety == tpb.TEMPLATE_VARIETY_QUOTA {
+					// only dispatch instances with a matching name
```
Comment (ozevren) [inline mixer/pkg/runtime/dispatcher/session.go]: This looks good, and the comment here is enough. But really should call out this behavior somewhere. Does the op-facing documentation needs to be updated?

**Pull request merged**